src/amg/azext_amg/migrate.py

```diff
@@ -0,0 +1,302 @@
+
+    for datasource in all_datasources:
+        if (datasource['name'], datasource['type']) not in restore_datasources_names:
+            if not dry_run:
+                create_datasource(restore_url, datasource, restore_headers)
+            datasources_created_summary.append(datasource['name'])
+        else:
+            if not dry_run:
+                logger.warning("Datasource %s already exists, remapping", datasource['name'])
+            datasources_remapped_summary.append(datasource['name'])
+    # grab all the new datasources now. since we have created the datasources, we can now do the mapping.
+    # do the mapping from the backup to the restore.
+    set_uid_mapping(all_datasources, all_restore_datasources)
+
+    return (datasources_created_summary, datasources_remapped_summary)
+
+
+def _migrate_folders(all_folders, all_restore_folders, restore_url, restore_headers, dry_run, overwrite):
+    folders_created_summary = []
+    folders_overwrote_summary = []
+
+    restore_folder_uids = {restore_content['uid'] for (restore_content, _) in all_restore_folders}
+    restore_folder_uids = set()
+
+    for folder in all_folders:
+        content_folder_settings, _ = folder
+
+        if content_folder_settings['uid'] in restore_folder_uids:
+            logger.warning("Folder %s already exists, skipping", content_folder_settings['title'])
+            continue
+
+        # create a folder if it does not exist
+        exists_before = check_folder_exists(restore_url, content_folder_settings, restore_headers)
+
+        is_successful = True
+        if not dry_run:
+            is_successful = create_folder(restore_url, content_folder_settings, restore_headers, overwrite)
+
+        if not is_successful:
+            continue
+
+        update_summary(exists_before,
+                       content_folder_settings['title'],
+                       folders_created_summary,
+                       folders_overwrote_summary)
+
+    return folders_created_summary, folders_overwrote_summary
+
+
+def _migrate_library_panels_and_dashboards(all_dashboards, all_library_panels_filtered, restore_url, restore_headers,
+                                           dry_run, overwrite):
+    library_panels_created_summary = {}
+    library_panels_overwrote_summary = {}
+    dashboards_created_summary = {}
+    dashboards_overwrote_summary = {}
+
+    # only update the library panels that are not included / excluded.
+    for library_panel in all_library_panels_filtered:
+        exists_before = check_library_panel_exists(restore_url, library_panel, restore_headers)
+
+        library_panel['id'] = None
+        is_successful = True
+        if not dry_run:
+            # in create_library_panel, it does the patch if the library panel already exists.
+            is_successful = create_library_panel(restore_url, library_panel, restore_headers, overwrite)
+
+        if not is_successful:
+            continue
```

Comment:
does 'migrate' have partial success behavior? Seems like if it fails to migrate over a library panel it will just skip on ahead instead of terminating

Reply:
Yeah, if it fails, it will skip over that library panel. If there happens to be an issue where the dashboard doesn't have the library panel, it will fail to create that dashboard to so it will also skip over it.